### PR TITLE
golang extension: added more APIs for RequestHeaderMap & ResponseHeaderMap interface

### PR DESCRIPTION
**Pull request by timebug**
Commit Message: added more APIs for RequestHeaderMap & ResponseHeaderMap interface for golang extension
Additional Description:

1. Added more APIs for golang header api:

```
type RequestHeaderMap interface {
	HeaderMap
	Protocol() string
	Scheme() string
	Method() string
	Host() string
	Path() string
}

type ResponseHeaderMap interface {
	HeaderMap
	Status() (int, bool)
}
```

2. Split trailer function to make the code more clearer.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by doujiang24 — COMMENTED** [review]
maybe we could introduce a new middle impl, i.e. `requestOrResponseHeaderMapIMpl`, to reduce these duplicated codes?
```
func (h *requestHeaderMapImpl) Set(key, value string)
func (h *responseHeaderMapImpl) Set(key, value string)
```

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 1]
```diff
@@ -81,42 +90,128 @@ func (h *httpHeaderMap) Values(key string) []string {
...
+	}
+	if h.headers != nil {
+		h.headers[key] = []string{value}
+	}
+
+	cAPI.HttpSetHeader(unsafe.Pointer(h.request.req), &key, &value, false)
+}
+
+func (h *responseHeaderMapImpl) Add(key, value string) {
+	if h.headers != nil {
```
[the remark this answers is not in the document]
Comment (timebug) [reply contrib/golang/filters/http/source/go/pkg/http/type.go]: fixed.

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 2]
```diff
@@ -81,42 +90,128 @@ func (h *httpHeaderMap) Values(key string) []string {
...
-	} else {
-		cAPI.HttpSetHeader(unsafe.Pointer(h.request.req), &key, &value, true)
+	cAPI.HttpSetHeader(unsafe.Pointer(h.request.req), &key, &value, true)
+}
+
+func (h *requestHeaderMapImpl) Del(key string) {
+	// Get all header values first before removing a key, since the del operation may not take affects immediately
+	// when it's invoked in a Go thread, instead, it will post a callback to run in the envoy worker thread.
+	// Otherwise, we may get outdated values in a following Get call.
+	if h.headers != nil {
```
[the remark this answers is not in the document]
Comment (timebug) [reply contrib/golang/filters/http/source/go/pkg/http/type.go]: fixed.

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 3]
```diff
@@ -32,32 +32,45 @@ const (
...
 }
 
-var _ api.HeaderMap = (*httpHeaderMap)(nil)
-
-func (h *httpHeaderMap) GetRaw(key string) string {
-	if h.isTrailer {
-		panic("unsupported yet")
+func (h *headerMapImpl) Range(f func(key, value string) bool) {
+	if h.headers == nil {
+		h.headers = cAPI.HttpCopyHeaders(unsafe.Pointer(h.request.req), h.headerNum, h.headerBytes)
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:45]: this is only for headers, should move to `requestOrResponseHeaderMapImpl`.

Comment (timebug) [reply contrib/golang/filters/http/source/go/pkg/http/type.go]: done.

**Review by doujiang24 — DISMISSED** [review]
LGTM.

@wbpcode could you please take a look? thank!

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 4]
```diff
@@ -81,53 +77,49 @@ func (h *httpHeaderMap) Values(key string) []string {
...
-	} else {
-		cAPI.HttpSetHeader(unsafe.Pointer(h.request.req), &key, &value, false)
-	}
+
+	cAPI.HttpSetHeader(unsafe.Pointer(h.request.req), &key, &value, false)
 }
 
-func (h *httpHeaderMap) Add(key, value string) {
+func (h *requestOrResponseHeaderMapImpl) Add(key, value string) {
+	if h.headers == nil {
```
Comment (wangfakang) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:95]: nit: So many identical checks for the header, so we can try to separate it to the header initialization phase. same elseeverywhere.

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 5]
```diff
@@ -137,11 +129,134 @@ func (h *httpHeaderMap) Range(f func(key, value string) bool) {
...
+}
+
+var _ api.ResponseHeaderMap = (*responseHeaderMapImpl)(nil)
+
+func (h *responseHeaderMapImpl) Status() int {
+	if str, ok := h.Get(":status"); ok {
+		v, _ := strconv.Atoi(str)
+		return v
+	}
+	return 0
```
Comment (wangfakang) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:176]: 0 means get status fail? It also has another meaning similar to the 499 code. such as below: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#format-strings
```
HTTP response code. Note that a response code of ‘0’ means that the server never sent the beginning of a response. This generally means that the (downstream) client disconnected.
```

Comment (timebug) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:176]: Yes, 0 means not found status in response, I see envoy Internal implementation also seem to take 0 as the default:

* https://github.com/envoyproxy/envoy/blob/main/source/common/http/utility.cc#L593
* https://github.com/envoyproxy/envoy/blob/main/source/common/formatter/substitution_formatter.cc#L1026

Do you think it's better to define it as `Status() (int, bool)`, if get fail, `return 0, false`?

Comment (wangfakang) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:176]: yeah

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**